### PR TITLE
[Feat] Add Loops Restart

### LINKED ISSUE #8126: ModelCheckpoint does not save checkpoint on training end
## 🚀 Feature
See title

### Motivation
When finishing training, either through keyboard interrupt, unexpected error, reaching the end of the intended training period, or any other means, it is very desirable to keep a checkpoint of the most recent state of our training.

### Pitch
Imagine you need to interrupt the current training, but the last checkpoint was made hours ago, and you cannot wait for the next checkpoint to be saved in 3000 more steps. You need the system to drop a checkpoint for you when you stop the training.

### Alternatives
Users could extend their own ModelCheckpoint to have an on_fit_end hook

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Fixes #8126.

TODOS:
[x] Add restart for training epoch loop
[x] Add restart for training batch loop
[x] Add restart for multiple optimizers
[x] Add support random samplers
[x] Add support `accumulate_grad_batches` (saving gradients)
[x] Add checkpoint creation on failure
[x] Add mechanism to prevent DDP deadlock on failure: Merged https://github.com/PyTorchLightning/pytorch-lightning/pull/8167
[x] Add restart for validation epoch loop
[x] Add restart for the validation batch loop
[x] Add restart for multiple validation dataloaders
[x] Add `val_check_internval`
[x] Add `FastForwardSampler` and `CaptureIterativeDataset` to recover dataset states. # https://github.com/PyTorchLightning/pytorch-lightning/pull/8307
[] Add support for `results / extras` on epoch_end for train
[] Add restore / restarting attribute function to Loops
[] Link progress tracking to Loop attributes
[] Investigate InfiniBatch for random seed (num_workers, ddp sampler)
[] Handle manual optimizer step (optimizer step / zero grad to take care of)
[] Test DistributedSampler
[] Add configuration mechanism to restart as currently supported. Start from next epoch.
[] Restart validation loop using `should_check_val`.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/progress.py [hunk 1]
```diff
@@ -28,10 +30,10 @@ class Tracker:
...
     Attributes set to ``None`` are treated as unused and are restricted.
     """
-    ready: Optional[int] = 0
-    started: Optional[int] = 0
-    processed: Optional[int] = 0
-    completed: Optional[int] = 0
+    ready: Optional[int] = field(default_factory=int)
+    started: Optional[int] = field(default_factory=int)
+    processed: Optional[int] = field(default_factory=int)
+    completed: Optional[int] = field(default_factory=int)
```
Comment (carmocca) [inline pytorch_lightning/trainer/progress.py:36]: This is not necessary. The reference issues you were experiencing were because they were classes. This doesn't happen to primitives (ints in this case)

Comment (tchaton) [reply pytorch_lightning/trainer/progress.py:36]: Thanks !

**Comment by tchaton** [comment]
This PR is closed for https://github.com/PyTorchLightning/pytorch-lightning/pull/8337
